src/Mvc/Mvc.RazorPages/src/DependencyInjection/PageConventionCollectionExtensions.cs

```diff
@@ -378,8 +446,18 @@ public static PageConventionCollection AuthorizeAreaFolder(this PageConventionCo
                 throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(folderPath));
             }
 
-            var authorizeFilter = new AuthorizeFilter(policy);
-            conventions.AddAreaFolderApplicationModelConvention(areaName, folderPath, model => model.Filters.Add(authorizeFilter));
+            var authorizeAttribute = new AuthorizeAttribute(policy);
+            conventions.AddAreaFolderApplicationModelConvention(areaName, folderPath, model =>
+            {
+                if (conventions.MvcOptions.EnableEndpointRouting)
+                {
+                    model.EndpointMetadata.Add(new AuthorizeAttribute(policy));
+                }
+                else
+                {
+                    model.Filters.Add(new AuthorizeFilter(policy));
+                }
+            });
```

Comment:
The constructor thing on the collection is gross, but this works pretty much like what I expected.